users/users.go

```diff
@@ -1,228 +1,440 @@
+
+    // Promote the user through ranks if enough days have passed
+    for {
+        nextRank := GetNextRank(currentRank)
+        if nextRank == nil {
+            log.Infof("User %s already has the highest rank '%s'", user.GetName(), user.RankName)
+            break
+        }
+
+        daysNeededForNextRank := nextRank.MinDays - currentRank.MinDays
+        log.Infof("Days needed to go from '%s' to '%s': %d days", currentRank.Name, nextRank.Name, daysNeededForNextRank)
+
+        if effectiveDays >= daysNeededForNextRank {
+            log.Infof("Promoting user %s from '%s' to '%s'", user.GetName(), currentRank.Name, nextRank.Name)
+            user.RankName = nextRank.Name
+            user.RankUpdatedAt = ptrTimeNow()
+            currentRank = nextRank
+            effectiveDays -= daysNeededForNextRank
+            promoted = true
+
+            log.Infof("After promotion, effective days remaining for user %s: %d", user.GetName(), effectiveDays)
+
+        } else {
+            break
+        }
+    }
+
+    // Save the user only if a promotion happened
+    if promoted {
+        if err := db.DBCon.Save(&user).Error; err != nil {
+            log.Errorf("Failed to save promoted user %s: %v", user.GetName(), err)
+            return err
+        }
+        log.Infof("User %s rank updated successfully", user.GetName())
+    } else {
+        log.Infof("No rank change needed for user %s", user.GetName())
+    }
+
+    return nil
+}
+
+
+
+func (user *User) EnsureRankUpdatedAtExists() error {
+    if user.RankUpdatedAt != nil {
+        return nil
+    }
+
+    // Try to find the latest rejoin event
+    var lastRejoin Event
+    err := db.DBCon.
+        Where("user_id = ? AND event = ?", user.ID, RejoinedGroupEventType).
+        Order("created_at DESC").
+        First(&lastRejoin).Error
+
+    if err == nil {
+        // Found a rejoin event -> use its date
+        user.RankUpdatedAt = &lastRejoin.CreatedAt
+        if user.RankName == "" {
+            user.RankName = Ranks[0].Name
+        }
+        if err := db.DBCon.Save(&user).Error; err != nil {
+            log.Errorf("Failed to save RankUpdatedAt after rejoin for user %s: %v", user.GetName(), err)
+            return err
+        }
+        log.Infof("Initialized RankUpdatedAt for user %s to rejoin date: %s", user.GetName(), lastRejoin.CreatedAt.Format("2006-01-02"))
+        return nil
+    }
+
+    // No rejoin found, fallback to first workout
+    var firstWorkout Workout
+    err = db.DBCon.
+        Where("user_id = ? AND deleted_at IS NULL", user.ID).
+        Order("created_at ASC").
+        First(&firstWorkout).Error
+
+    if err != nil {
+        log.Warnf("User %s has no workouts to set RankUpdatedAt", user.GetName())
+        return nil // No workouts either
+    }
+
+    user.RankUpdatedAt = &firstWorkout.CreatedAt
+    if user.RankName == "" {
+        user.RankName = Ranks[0].Name
+    }
+    if err := db.DBCon.Save(&user).Error; err != nil {
```

Comment:
better code would be

```
func (user *User) UpdateRank(rank string) error {}
```
and then use it here.

(every time you see dbcon.save there's probably a good place for that)